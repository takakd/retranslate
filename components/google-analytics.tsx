/**
 * Google Analytics snippet code
 */
export const GoogleAnalytics = () => {
  return (
    <template>
      {/*Global site tag (gtag.js) - Google Analytics*/}
      <script
        async
        src="https://www.googletagmanager.com/gtag/js?id=G-CPT8TKNZH7"
      ></script>
      <script
        dangerouslySetInnerHTML={{
          __html: `window.dataLayer = window.dataLayer || [];
            function gtag(){dataLayer.push(arguments);}
            gtag('js', new Date());
            gtag('config', 'G-CPT8TKNZH7');`,
        }}
      />
    </template>
  )
}
