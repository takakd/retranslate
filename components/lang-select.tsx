import React, { useEffect, useState } from 'react'

import { Dropdown, DropdownButton } from 'react-bootstrap'

import { LangType } from '../grpc/translator_pb'

/**
 * LangSelectChangeEvent is the event object of LangSelectProps.onLangChange event.
 */
export interface LangSelectChangeEvent {
  /**
   * Selected language type
   */
  lang: LangType
}

/**
 * LangSelect component props
 */
export interface LangSelectProps {
  /**
   * Language type
   */
  lang?: LangType

  /**
   * LangType change event handler.
   * @param {LangSelectChangeEvent} event event object.
   */
  onLangChange?: (event: LangSelectChangeEvent) => void

  /**
   * Language options
   */
  langOptions: {
    label: string
    value: LangType
  }[]
}

/**
 * LangSelect component
 *
 * @param {LangSelectProps} props
 */
export const LangSelect = (props: LangSelectProps) => {
  const [langLabel, setLangLabel] = useState<string>('')

  useEffect(() => {
    // Set option label by props.lang.
    setLangLabel(toLangLabel(props.lang))
  })

  /**
   * Creates the language label by LangType.
   * @param {LangType} lang language type.
   * @returns {string} language label.
   */
  const toLangLabel = (lang: LangType): string => {
    for (const option of props.langOptions) {
      if (lang === option.value) {
        return option.label
      }
    }
    return 'Language'
  }

  /**
   * Fired on clicking the language option.
   * @param {LangType} lang Selected language.
   */
  const onClick = (lang: LangType) => {
    if (!props.onLangChange) {
      return
    }
    props?.onLangChange({
      lang,
    })
  }

  return (
    <DropdownButton
      key={props.lang}
      title={langLabel}
      variant="secondary"
      renderMenuOnMount={true}
    >
      {props.langOptions.map((option) => {
        return (
          <Dropdown.Item
            as="button"
            key={option.value}
            onClick={() => onClick(option.value)}
          >
            {option.label}
          </Dropdown.Item>
        )
      })}
    </DropdownButton>
  )
}
