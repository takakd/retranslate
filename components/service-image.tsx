import React from 'react'

/**
 * Translation API service name
 */
export enum Service {
  AWS = 'aws',
  Google = 'google',
}

/**
 * ServiceImage component props
 */
export class ServiceImageProps {
  /**
   * Translation API service name
   */
  service: Service
}

/**
 * ServiceImage component
 *
 * @param {ServiceImageProps} props
 */
export const ServiceImage = (props: ServiceImageProps) => {
  if (props.service === Service.Google) {
    return <img src="/logo-google.png" alt="google" width="20" height="20" />
  } else if (props.service === Service.AWS) {
    return <img src="/logo-aws.png" alt="aws" width="28" height="20" />
  }
}
