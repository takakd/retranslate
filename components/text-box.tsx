import React, { FormEvent, useRef } from 'react'

import { CloseButton, Spinner } from 'react-bootstrap'
import { ArrowRight, Clipboard } from 'react-bootstrap-icons'

import { ServiceImage } from '../components/service-image'

/**
 * Translation service name
 */
export enum Service {
  AWS = 'aws',
  Google = 'google',
}

/**
 * TextBoxInputEvent represents an event object of TextBoxProps.onInput event.
 */
export interface TextBoxInputEvent {
  /**
   * Text in textarea
   */
  text: string
}

/**
 * Props of TextBox component
 */
export interface TextBoxProps {
  /**
   * Initial text in textarea
   */
  text: string
  /**
   * Translation service name
   */
  service?: Service
  /**
   * Input event handler.
   * @param {TextBoxInputEvent} event event object.
   */
  onInput?: (event: TextBoxInputEvent) => void
  /**
   * Clear text event handler.
   */
  onClearText?: () => void
  /**
   * Class name
   */
  className?: string
  /**
   * Class name prefix of child elements.
   */
  classNamePrefix?: string
  /**
   * Placeholder text in textarea.
   */
  placeHolder?: string

  /**
   * Show arrow on the left of textarea.
   */
  showArrow?: boolean
  /**
   * Show loading icon on the left of textarea.
   */
  loading?: boolean
}

/**
 * TextBox component.
 *
 * @param {TextBoxProps} props
 */
export const TextBox = (props: TextBoxProps) => {
  const textareaRef = useRef(null)
  const className = props?.className ? props.className : 'text-box'
  const classNamePrefix = props?.classNamePrefix
    ? props.classNamePrefix
    : 'text-box'

  /**
   * Fired on inputting texts in textarea.
   * @param {React.FormEvent} event
   */
  const onInput = (event: FormEvent) => {
    if (!props.onInput) {
      return
    }

    const target = event.target as HTMLTextAreaElement
    props.onInput({
      text: target.value,
    })
  }

  /**
   * Fired on clicking the clipboard button.
   */
  const onClickClipboard = () => {
    // Copy text in textarea to the OS clipboard.
    textareaRef.current.select()
    document.execCommand('copy')
  }

  return (
    <div className={`d-flex flex-row align-items-center w-100 ${className}`}>
      {
        // Show arrow icon.
        props.showArrow && !props.loading && (
          <div className="icon">
            <ArrowRight size={20} />
          </div>
        )
      }
      {
        // Show loading icon when loading.
        props.loading && (
          <div className="icon">
            <Spinner animation="border" size="sm" />
          </div>
        )
      }
      <div className={`d-flex flex-column ${classNamePrefix}-text-container`}>
        {
          <div className={`${className}-header`}>
            {
              // Show service icon
              props.service && <ServiceImage service={props.service} />
            }
          </div>
        }
        <div
          className={`position-relative w-100 ${classNamePrefix}-textarea-container`}
        >
          <textarea
            ref={textareaRef}
            className={`h-100 ${classNamePrefix}-textarea`}
            placeholder={props?.placeHolder}
            defaultValue={props.text}
            onInput={(event) => onInput(event)}
          />
          {
            // Show erase button when text is entered and the event handler exists.
            props.text !== '' && props?.onClearText && (
              <CloseButton
                className={`position-absolute close-button p-1 ${classNamePrefix}-close-button`}
                onClick={props?.onClearText}
              />
            )
          }
          {
            // Show clipboard button when text is entered.
            props.text !== '' && (
              <Clipboard
                className={`position-absolute ${classNamePrefix}-clipboard-button`}
                onClick={onClickClipboard}
              />
            )
          }
        </div>
      </div>

      <style jsx>{`
        .${className}-header {
          min-height: 30px;
        }

        .${classNamePrefix}-textarea {
          width: 100%;
          resize: none;
          padding: 0.5rem 0.5rem 36px 0.5rem;
        }
        .${classNamePrefix}-text-container,
          .${classNamePrefix}-textarea-container {
          flex: 1;
        }
        .icon {
          min-width: 30px;
          text-align: center;
        }
        ::placeholder {
          color: #aaa;
          opacity: 1;
        }
      `}</style>
      <style jsx global>{`
        .${classNamePrefix}-close-button {
          top: 1px;
          right: 6px;
        }
        .${classNamePrefix}-clipboard-button {
          cursor: pointer;
          bottom: 10px;
          right: 10px;
        }
      `}</style>
    </div>
  )
}
