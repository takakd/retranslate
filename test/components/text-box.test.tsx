import React from 'react'

import { Service, TextBox, TextBoxProps } from '../../components/text-box'
import { fireEvent, render } from '../testUtils'

describe('TextBox', () => {
  it('matches snapshot', () => {
    const props: TextBoxProps = {
      text: 'test',
      service: Service.Google,
      placeHolder: 'placeholder',
      showArrow: true,
      loading: true,
    }

    const { asFragment } = render(<TextBox {...props} />)

    expect(asFragment()).toMatchSnapshot()
  })

  it('on input', () => {
    const props: TextBoxProps = {
      text: 'test',
      service: Service.Google,
      onInput: jest.fn(),
    }

    const { container } = render(<TextBox {...props} />)

    fireEvent.input(container.getElementsByTagName('textarea')[0])

    expect(props.onInput).toBeCalledTimes(1)
    expect(props.onInput).toBeCalledWith({ text: props.text })
  })

  it('on close', async () => {
    const props: TextBoxProps = {
      text: 'test',
      service: Service.Google,
      onClearText: jest.fn(),
    }

    const { container } = render(<TextBox {...props} />)

    fireEvent.click(container.getElementsByClassName('close-button')[0])

    expect(props.onClearText).toBeCalledTimes(1)
  })

  it('on clipboard', async () => {
    const props: TextBoxProps = {
      text: 'test',
      service: Service.Google,
      classNamePrefix: 'test',
    }

    const { container } = render(<TextBox {...props} />)

    document.execCommand = jest.fn()

    fireEvent.click(
      container.getElementsByClassName('test-clipboard-button')[0]
    )

    expect(document.execCommand).toBeCalledWith('copy')
    expect(document.execCommand).toBeCalledTimes(1)
  })
})
