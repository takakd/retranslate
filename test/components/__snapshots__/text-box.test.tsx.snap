// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextBox matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="jsx-38414541 jsx-475603417 d-flex flex-row align-items-center w-100 text-box"
  >
    <div
      class="jsx-38414541 jsx-475603417 icon"
    >
      <div
        class="spinner-border spinner-border-sm"
      />
    </div>
    <div
      class="jsx-38414541 jsx-475603417 d-flex flex-column text-box-text-container"
    >
      <div
        class="jsx-38414541 jsx-475603417 text-box-header"
      >
        <img
          alt="google"
          height="20"
          src="/logo-google.png"
          width="20"
        />
      </div>
      <div
        class="jsx-38414541 jsx-475603417 position-relative w-100 text-box-textarea-container"
      >
        <textarea
          class="jsx-38414541 jsx-475603417 h-100 text-box-textarea"
          placeholder="placeholder"
        >
          test
        </textarea>
        <svg
          class="position-absolute text-box-clipboard-button"
          fill="currentColor"
          height="1em"
          viewBox="0 0 16 16"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M4 1.5H3a2 2 0 0 0-2 2V14a2 2 0 0 0 2 2h10a2 2 0 0 0 2-2V3.5a2 2 0 0 0-2-2h-1v1h1a1 1 0 0 1 1 1V14a1 1 0 0 1-1 1H3a1 1 0 0 1-1-1V3.5a1 1 0 0 1 1-1h1v-1z"
          />
          <path
            d="M9.5 1a.5.5 0 0 1 .5.5v1a.5.5 0 0 1-.5.5h-3a.5.5 0 0 1-.5-.5v-1a.5.5 0 0 1 .5-.5h3zm-3-1A1.5 1.5 0 0 0 5 1.5v1A1.5 1.5 0 0 0 6.5 4h3A1.5 1.5 0 0 0 11 2.5v-1A1.5 1.5 0 0 0 9.5 0h-3z"
          />
        </svg>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
